File: Linked List/merge_two_sorted_lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution: #O(n) time complexity, #O(1) space complexity
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        curr = dummy

        while list1 and list2:
            if list1.val > list2.val:
                curr.next = list2
                list2 = list2.next
            else:
                curr.next = list1
                list1 = list1.next
            curr = curr.next

        if list1:
            curr.next = list1
        else:
            curr.next = list2

        return dummy.next


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class MySolution: #Same complexity as above
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        l1 = list1
        l2 = list2
        dummy = ListNode()
        cur = dummy

        while l1 and l2:
            if l1.val <= l2.val:
                temp = l1.next
                cur.next = l1
                cur.next.next = None
                l1 = temp
            else:
                temp = l2.next
                cur.next = l2
                cur.next.next = None
                l2 = temp

            cur = cur.next

        if l1:
            cur.next = l1
        
        if l2:
            cur.next = l2

        return dummy.next

File: Linked List/merge_two_sorted_lists.js
/*
https://leetcode.com/problems/merge-two-sorted-lists/description/
*/


/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} list1
     * @param {ListNode} list2
     * @return {ListNode}
     * Time complexity O(n+m), Space complexity O(1)
     */
    mergeTwoLists(list1, list2) {
        const dummy = new ListNode()
        let cur = dummy

        while(list1 !== null && list2 !== null){
            if(list1.val <= list2.val){
                let temp = list1
                list1 = list1.next
                cur.next = temp
                cur = cur.next 
            }else{
                let temp = list2
                list2 = list2.next
                cur.next = temp
                cur = cur.next 
            }
        }

        if (list1 !== null){
            cur.next = list1
        }

        if(list2 !== null){
            cur.next = list2
        }

        return dummy.next

    }
}
